ci-operator/step-registry/osd-ccs/cluster/private/osd-ccs-cluster-private-commands.sh

```diff
@@ -0,0 +1,109 @@
+  if [ "${api_lisening}" != "internal" ]; then
+    logger "ERROR" "Cluster API Listening is '${api_lisening}' unexpectedly"
+    return 1
+  else
+    logger "INFO" "check_private: Cluster API Listening is '${api_lisening}'"
+  fi
+
+  # 2. check on GCP the DNS record-sets
+  tmp_output=$(mktemp)
+  gcloud_auth
+
+  # find out the private DNS zone name and public DNS zone name
+  priv_zone_name=$(gcloud dns managed-zones list --filter="name~${CLUSTER_NAME}" --format='value(name)')
+  priv_zone_dns=$(gcloud dns managed-zones list --filter="name~${CLUSTER_NAME}" --format='value(dnsName)')
+  pub_zone_dns="${priv_zone_dns#*.}"
+  pub_zone_name=$(gcloud dns managed-zones list --filter="dnsName=${pub_zone_dns}" --format='value(name)')
+  logger "INFO" "The DNS private zone: ${priv_zone_name} (${priv_zone_dns})"
+  logger "INFO" "The DNS public zone: ${pub_zone_name} (${pub_zone_dns})"
+
+  # list DNS record-sets
+  cmd="gcloud dns record-sets list --zone ${pub_zone_name} --filter=\"(type=A OR type=SRV) AND NOT name~rh-api\" 2>/dev/null | tee ${tmp_output}"
+  logger "INFO" "Running Command '${cmd}'"
+  eval "${cmd}"
+  num_priv_ip=$(grep -P "\s+10\.[0-9\.]+$" "${tmp_output}" | wc -l)
+  logger "INFO" "There are ${num_priv_ip} dns record-sets in public zone '${pub_zone_name}' with private IP addresses."
+  # Expecting both api and *.apps dns record-sets in public zone are with private IP addresses
+  if [ ${num_priv_ip} -lt 2 ]; then
+    logger "ERROR" "One or more api/*.apps dns record-sets in zone '${pub_zone_name}' are with non-private IP address."
+    return 1
+  fi
+  cmd="gcloud dns record-sets list --zone ${priv_zone_name} --filter=\"(type=A OR type=SRV) AND NOT name~rh-api\" 2>/dev/null | tee ${tmp_output}"
+  logger "INFO" "Running Command '${cmd}'"
+  eval "${cmd}"
+  num_priv_ip=$(grep -P "\s+10\.[0-9\.]+$" "${tmp_output}" | wc -l)
+  logger "INFO" "There are ${num_priv_ip} dns record-sets in zone '${priv_zone_name}' with private IP addresses."
+  # Expecting all api, api-int and *.apps dns record-sets in private zone are with private IP addresses
+  if [ ${num_priv_ip} -lt 3 ]; then
+    logger "ERROR" "One or more api/api-int/*.apps dns record-sets in private zone '${priv_zone_name}' are with non-private IP address."
+    return 1
+  fi
+
+  logger "INFO" "check_private: Cluster DNS record-sets check passed"
+  return 0
+}
+
+trap 'CHILDREN=$(jobs -p); if test -n "${CHILDREN}"; then kill ${CHILDREN} && wait; fi' TERM
+
+# Log in
+OCM_VERSION=$(ocm version)
+OCM_TOKEN=$(cat "${CLUSTER_PROFILE_DIR}/ocm-token")
+logger "INFO" "Logging into ${OCM_LOGIN_ENV} with offline token using ocm cli ${OCM_VERSION}"
+ocm login --url "${OCM_LOGIN_ENV}" --token "${OCM_TOKEN}"
+
+# Required
+GCP_CREDENTIALS_FILE="${CLUSTER_PROFILE_DIR}/osd-ccs-gcp.json"
+
+if [ -z "${CLUSTER_NAME}" ]; then
+    CLUSTER_NAME=$(cat "${SHARED_DIR}/cluster-name")
+fi
+CLUSTER_ID=$(cat "${SHARED_DIR}/cluster-id")
+
+set +o errexit
```

Comment:
Do we do not want the job exit as failure when `check_private` return non-zero ?

Reply:
Good question, let me confirm. Thanks!

Reply:
The function `check_private` will be executed fully, and its return code would be the exit code of the step, so if the function returns non-zero, it's true the step will exit with non-zero code.

Reply:
if that, what is the purpose of this line, why not drop it?

Reply:
@jianlinliu FYI my local testing results of the logic, where if without the line the function `check_private` may not execute fully (no any log/output), it'll exit if the `grep` failed. 

**Scenario A: without `set +o errexit`**
```
[jiwei@jiwei-thinkpadt14gen4 ~]$ cat osd-ccs-cluster-private-commands.sh 
#!/bin/bash
set -o nounset
set -o errexit
set -o pipefail

tmp_output1="/tmp/pub"
tmp_output2="/tmp/priv"

function logger() {
  local -r log_level=$1; shift
  local -r log_msg=$1; shift
  echo "$(date -u --rfc-3339=seconds) - ${log_level}: ${log_msg}"
}

# Post installation check for private cluster
function check_private() {
  local num_priv_ip

  num_priv_ip=$(grep -P "\s+10\.[0-9\.]+$" "${tmp_output1}" | wc -l)
  logger "INFO" "There are ${num_priv_ip} dns record-sets in public zone with private IP addresses."
  # Expecting both api and *.apps dns record-sets in public zone are with private IP addresses
  if [ ${num_priv_ip} -lt 2 ]; then
    logger "ERROR" "One or more api/*.apps dns record-sets in public zone are with non-private IP address."
    return 1
  fi
  num_priv_ip=$(grep -P "\s+10\.[0-9\.]+$" "${tmp_output2}" | wc -l)
  logger "INFO" "There are ${num_priv_ip} dns record-sets in private zone with private IP addresses."
  # Expecting all api, api-int and *.apps dns record-sets in private zone are with private IP addresses
  if [ ${num_priv_ip} -lt 3 ]; then
    logger "ERROR" "One or more api/api-int/*.apps dns record-sets in private zone are with non-private IP address."
    return 1
  fi

  logger "INFO" "check_private: Cluster DNS record-sets check passed"
  return 0
}

#set +o errexit
check_private
[jiwei@jiwei-thinkpadt14gen4 ~]$ ./osd-ccs-cluster-private-commands.sh 
[jiwei@jiwei-thinkpadt14gen4 ~]$ echo $?
1
[jiwei@jiwei-thinkpadt14gen4 ~]$ 
```
**Scenario B: with `set +o errexit`**
```
[jiwei@jiwei-thinkpadt14gen4 ~]$ cat osd-ccs-cluster-private-commands.sh 
#!/bin/bash
set -o nounset
set -o errexit
set -o pipefail

tmp_output1="/tmp/pub"
tmp_output2="/tmp/priv"

function logger() {
  local -r log_level=$1; shift
  local -r log_msg=$1; shift
  echo "$(date -u --rfc-3339=seconds) - ${log_level}: ${log_msg}"
}

# Post installation check for private cluster
function check_private() {
  local num_priv_ip

  num_priv_ip=$(grep -P "\s+10\.[0-9\.]+$" "${tmp_output1}" | wc -l)
  logger "INFO" "There are ${num_priv_ip} dns record-sets in public zone with private IP addresses."
  # Expecting both api and *.apps dns record-sets in public zone are with private IP addresses
  if [ ${num_priv_ip} -lt 2 ]; then
    logger "ERROR" "One or more api/*.apps dns record-sets in public zone are with non-private IP address."
    return 1
  fi
  num_priv_ip=$(grep -P "\s+10\.[0-9\.]+$" "${tmp_output2}" | wc -l)
  logger "INFO" "There are ${num_priv_ip} dns record-sets in private zone with private IP addresses."
  # Expecting all api, api-int and *.apps dns record-sets in private zone are with private IP addresses
  if [ ${num_priv_ip} -lt 3 ]; then
    logger "ERROR" "One or more api/api-int/*.apps dns record-sets in private zone are with non-private IP address."
    return 1
  fi

  logger "INFO" "check_private: Cluster DNS record-sets check passed"
  return 0
}

set +o errexit
check_private
[jiwei@jiwei-thinkpadt14gen4 ~]$ ./osd-ccs-cluster-private-commands.sh 
2024-04-03 05:36:13+00:00 - INFO: There are 0 dns record-sets in public zone with private IP addresses.
2024-04-03 05:36:13+00:00 - ERROR: One or more api/*.apps dns record-sets in public zone are with non-private IP address.
[jiwei@jiwei-thinkpadt14gen4 ~]$ echo $?
1
[jiwei@jiwei-thinkpadt14gen4 ~]$ 
```
**The input files content:** 
```
[jiwei@jiwei-thinkpadt14gen4 ~]$ cat /tmp/pub /tmp/priv
NAME                                                          TYPE  TTL  DATA
api.ci-op-2fvrkd1p-1c8ac.qe.gcp.devcluster.openshift.com.     A     60   34.67.23.216
*.apps.ci-op-2fvrkd1p-1c8ac.qe.gcp.devcluster.openshift.com.  A     30   34.171.141.22

NAME                                                           TYPE  TTL  DATA
api.ci-op-2fvrkd1p-1c8ac.qe.gcp.devcluster.openshift.com.      A     60   10.0.0.3
api-int.ci-op-2fvrkd1p-1c8ac.qe.gcp.devcluster.openshift.com.  A     60   10.0.0.3
*.apps.ci-op-2fvrkd1p-1c8ac.qe.gcp.devcluster.openshift.com.   A     30   34.171.141.22

[jiwei@jiwei-thinkpadt14gen4 ~]$ ```

Reply:
That sounds some line in the script return non-zero, which hit `errexit` when the option is enabled. 

So I think we need to locate which line cause that, and enhance the script to ensure it exit as expectation but not have to disable `errexit` option. Maybe you can `set -x` in the script to get the debug info to locate which line return non-zero value when there is no this line.